python_modules/libraries/dagster-dbt/dagster_dbt/dbt_artifacts.py

```diff
@@ -0,0 +1,152 @@
+    def __init__(
+        self,
+        project_dir: Union[Path, str],
+        *,
+        target_folder: Union[Path, str] = "target",
+        prepare_command: List[str] = ["parse", "--quiet"],
+        package_data_dir: Optional[Union[Path, str]] = None,
+    ):
+        """A utility class to help manage dbt artifacts in different deployment contexts.
+
+        This class provides a setup to solve for these goals:
+        * During development, reload the manifest at run time to pick up any changes.
+        * When deploying, expect a manifest that was prepared at build time to reduce start-up time.
+        * Handle the scenario when the dbt project is copied in to a directory for packaging.
+
+        Args:
+            project_dir (Union[str, Path]):
+                The directory of the dbt project.
+            target_folder (Union[str, Path]):
+                The folder in the project project directory to output artifacts.
+                Default: "target"
+            prepare_command: The dbt cli command to run to prepare the manifest.json
+                Default: ["parse", "--quiet"]
+            package_data_dir (Union[str, Path]):
+                A directory that will contain a copy of the dbt project, synced at build time.
+        """
+        self.project_dir = Path(project_dir)
+        if not self.project_dir.exists():
+            raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
+        self._target_folder = Path(target_folder)
+        self._prepare_command = prepare_command
+        # this is ok if it doesn't exist, will get created by `prepare`
+        self._package_data_dir = Path(package_data_dir) if package_data_dir else None
+
+    @public
+    @cached_property  # to prevent repeat compilation
+    def manifest_path(self) -> Path:
+        """The path to the manifest.json, compiling the manifest first if in development or
+        ensuring it already exists if not.
+        """
+        if self._should_prepare_at_runtime():
+            return self._prepare_manifest()
+
+        p = self._base_dir.joinpath(self._target_folder, "manifest.json")
+        if not p.exists():
+            raise DagsterDbtManifestNotPreparedError(
+                f"Did not find prepared manifest.json at expected path {p}.\n"
+                "If this is an environment that is expected to prepare the manifest at run time, "
+                "set the environment variable DAGSTER_DBT_PARSE_PROJECT_ON_LOAD."
+            )
+        return p
+
+    @public
+    def prepare(self, *, quiet=False) -> None:
+        """A method that can be called as part of the deployment process to handle
+        preparing the manifest and if package_data_dir is set, handle copying
+        the dbt project to that directory.
+
+        Args:
+            quiet (bool):
+                Disable logging
+                Default: False
+        """
+        logger = None if quiet else logging.getLogger("dagster-dbt-artifacts")
```

Comment:
is there not some sort of `/dev/null` style logger we can use here so you don't have to check against falsey everywhere?

Reply:
didn't love what i saw from the fist hit in search https://stackoverflow.com/questions/13521981/implementing-an-optional-logger-in-code

will re-evaluate

Reply:
Yeah not a big deal. Linked solutions probably more trouble than they are worth.